bundles/org.eclipse.ui.ide.application/src/org/eclipse/ui/internal/ide/application/DelayedEventsProcessor.java

```diff
@@ -183,90 +198,119 @@ public void run() {
-								details.path);
-						MessageDialog.open(MessageDialog.ERROR, window.getShell(),
-								IDEWorkbenchMessages.OpenDelayedFileAction_title, msg, SWT.SHEET);
-					}
-					try {
-						IEditorPart openEditor = IDE.openInternalEditorOnFileStore(page, details.fileStore);
-						Shell shell = window.getShell();
-						if (shell != null) {
-							if (shell.getMinimized())
-								shell.setMinimized(false);
-							shell.forceActive();
-						}
-
-						if (details.line >= 1) {
-							try {
+					return;
+				}
+				IWorkbenchPage page = window.getActivePage();
+				if (page == null) {
+					String msg = NLS.bind(IDEWorkbenchMessages.OpenDelayedFileAction_message_noWindow,
+							details.path);
+					MessageDialog.open(MessageDialog.ERROR, window.getShell(),
+							IDEWorkbenchMessages.OpenDelayedFileAction_title, msg, SWT.SHEET);
+					return;
+				}
+				Shell shell = window.getShell();
+				if (shell != null) {
+					if (shell.getMinimized())
+						shell.setMinimized(false);
+					shell.forceActive();
+				}
+				try {
+					if (details.fileInfo.isDirectory()) {
+						// Open a directory
+						File directory = new File(details.fileStore.toURI());
+						for (IProject project : ResourcesPlugin.getWorkspace().getRoot().getProjects()) {
+							IPath location = project.getLocation();
+							if (location != null && directory.equals(location.toFile())) {
 								/*
-								 * Do things with reflection to avoid having to rely on the text editor
-								 * plug-ins.
+								 * The directory is the root of an existing project.
+								 * Select in Project Explorer.
+								 * If open, then refresh, else open.
 								 */
-								Bundle textEditorBundle = Platform.getBundle(TEXTEDITOR_BUNDLE_NAME);
-								if (textEditorBundle != null) {
-									Class<?> textEditorClass = textEditorBundle.loadClass(TEXTEDITOR_CLASS_NAME);
-									if (textEditorClass != null) {
-										Object textEditor = invoke(openEditor, "getAdapter", //$NON-NLS-1$
-												new Class[] { Class.class }, new Object[] { textEditorClass });
-										if (textEditor != null) {
-											openEditor = (IEditorPart) textEditor;
-										}
+								IViewPart view = page.showView(ProjectExplorer.VIEW_ID);
+								if (view instanceof ISetSelectionTarget setSelectionTarget) {
+									setSelectionTarget.selectReveal(new StructuredSelection(project));
+									ICommandService commandService = view.getViewSite().getService(ICommandService.class);
+									Command command = commandService.getCommand(project.isOpen() ?
+										IWorkbenchCommandConstants.FILE_REFRESH :
+										IWorkbenchCommandConstants.PROJECT_OPEN_PROJECT);
+									if (command != null && command.isHandled() && command.isEnabled()) {
+										command.executeWithChecks(new ExecutionEvent());
 									}
 								}
+								return;
+							}
+						}
+						// Fall back to import
+						SmartImportWizard wizard = new SmartImportWizard();
```

Comment:
Starting the smart import wizard may be refactored out into a private method to avoid code duplication

Reply:
Can you please elaborate on the deduplication? I'm not seeing the SmartImportWizard used anywhere else in this file. It seems to be used in OpenFolderCommand and ConfigureProjectHandler but slightly differently.

Reply:
> Can you please elaborate on the deduplication? I'm not seeing the SmartImportWizard used anywhere else in this file. It seems to be used in OpenFolderCommand and ConfigureProjectHandler but slightly differently.

Sorry. I now looked again. There's no code duplication. I mis-read this.